src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -2174,9 +2405,218 @@ internal static bool CheckValEscape(SyntaxNode node, BoundExpression expr, uint
 
                     return CheckValEscape(binary.Left.Syntax, binary.Left, escapeFrom, escapeTo, checkingReceiver: false, diagnostics: diagnostics) &&
                            CheckValEscape(binary.Right.Syntax, binary.Right, escapeFrom, escapeTo, checkingReceiver: false, diagnostics: diagnostics);
+
+                case BoundKind.UserDefinedConditionalLogicalOperator:
+                    var uo = (BoundUserDefinedConditionalLogicalOperator)expr;
+
+                    return CheckValEscape(uo.Left.Syntax, uo.Left, escapeFrom, escapeTo, checkingReceiver: false, diagnostics: diagnostics) &&
+                           CheckValEscape(uo.Right.Syntax, uo.Right, escapeFrom, escapeTo, checkingReceiver: false, diagnostics: diagnostics);
+
+                case BoundKind.QueryClause:
+                    var clauseValue = ((BoundQueryClause)expr).Value;
+                    return CheckValEscape(clauseValue.Syntax, clauseValue, escapeFrom, escapeTo, checkingReceiver: false, diagnostics: diagnostics);
+
+                case BoundKind.RangeVariable:  
+                    var variableValue = ((BoundRangeVariable)expr).Value;
+                    return CheckValEscape(variableValue.Syntax, variableValue, escapeFrom, escapeTo, checkingReceiver: false, diagnostics: diagnostics);
+
+                case BoundKind.ObjectInitializerExpression:
+                    var initExpr = (BoundObjectInitializerExpression)expr;
+                    return CheckValEscapeOfObjectInitializer(initExpr, escapeFrom, escapeTo, diagnostics);
+
+                // this would be correct implementation for CollectionInitializerExpression 
+                // however it is unclear if it is reachable since the initialized type must implement IEnumerable
+                case BoundKind.CollectionInitializerExpression:
+                    var colExpr = (BoundCollectionInitializerExpression)expr;
+                    return CheckValEscape(colExpr.Initializers, escapeFrom, escapeTo, diagnostics);
+
+                // this would be correct implementation for CollectionElementInitializer 
+                // however it is unclear if it is reachable since the initialized type must implement IEnumerable
+                case BoundKind.CollectionElementInitializer:
+                    var colElement = (BoundCollectionElementInitializer)expr;
+                    return CheckValEscape(colElement.Arguments, escapeFrom, escapeTo, diagnostics);
+
+                default:
+                    throw ExceptionUtilities.UnexpectedValue($"{expr.Kind} expression of {expr.Type} type");
+
+                #region "cannot produce ref-like values"
+//                case BoundKind.ThrowExpression:
```

Comment:
I'm worried about those comments. Could we just un-comment and stick on the default case?

Reply:
I was thinking about making this an assert and return some default value - like "External". 
I am afraid it could lead to a node not handled correctly without our knowledge and setting up a compat precedent. That pushed me over to make it a throw. We can discuss.

Note - you have to be dealing with ref-like types to hit this, so impact is limited to a realatively small circle or users of some unexpected feature in combination with ref-like types. 
I'd rather get a bug report here than a silent behavior.


---
In reply to: [141158529](https://github.com/dotnet/roslyn/pull/22303#discussion_r141158529) [](ancestors = 141158529)

Reply:
I agree though very soo the comments will no longer serve the purpose. For now, i'd keep them so we do not need to redo the exercise of proving that all nodes are accounted for. Long term we should delete them.

---
In reply to: [141165552](https://github.com/dotnet/roslyn/pull/22303#discussion_r141165552) [](ancestors = 141165552,141158529)